source/common/upstream/subset_lb.cc

```diff
@@ -101,6 +111,72 @@ SubsetLoadBalancer::~SubsetLoadBalancer() {
   });
 }
 
+void SubsetLoadBalancer::rebuildSingle() {
+  if (single_key_.empty()) {
+    return;
+  }
+
+  // Because PriorityUpdateCb doesn't give a modified list (only added and removed), it is
+  // faster to just rebuild this map than try to figure out if any hosts had their metadata
+  // changed, and then figure out the old and new value (to remove from the old key in this map
+  // and insert in the new key).
+  single_host_per_subset_map_.clear();
+
+  uint32_t collision_count = 0;
+  for (const auto& host_set : original_priority_set_.hostSetsPerPriority()) {
+    for (const auto& host : host_set->hosts()) {
+      MetadataConstSharedPtr metadata = host->metadata();
+      const auto& filter_metadata = metadata->filter_metadata();
+      auto filter_it = filter_metadata.find(Config::MetadataFilters::get().ENVOY_LB);
+      if (filter_it != filter_metadata.end()) {
+        const auto& fields = filter_it->second.fields();
+        auto fields_it = fields.find(single_key_);
+        if (fields_it != fields.end()) {
+          auto [iterator, did_insert] =
+              single_host_per_subset_map_.try_emplace(fields_it->second, host);
+          if (!did_insert) {
+            // Two hosts with the same metadata value were found. Ignore all but one of them, and
+            // set a metric for how many times this happened.
+            collision_count++;
+          }
+        }
+      }
+    }
+  }
+
+  // This stat isn't added to `ClusterStats` because it wouldn't be used
+  // for nearly all clusters, and is only set during configuration updates,
+  // not in the data path, so performance of looking up the stat isn't critical.
+  Stats::StatNameManagedStorage name_storage("lb_subsets_single_host_per_subset_duplicate",
```

Comment:
Might not matter but maybe we'd want to just cache the resulting gauge in a `Stats::Gauge*` member variable to avoid the stat lookup? 

Maybe @jmarantz has thoughts here?

Reply:
Yeah, that might make sense.

Another change I considered (outside the scope of this change), would be to break up `ALL_CLUSTER_STATS`, so that stats specific to a certain LB are created by that LB. Right now we publish ~6 stats for subset_lb even if the cluster doesn't have any subset config. Similar for `lb_zone`. If those were a separate set that's only created by subset_lb, we could just add this stat to that set and not treat it specially.

Reply:
I think it's as simple as storing the `Gauge&` in the structure (e.g. right after `store_`).